src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -0,0 +1,223 @@
+    // Actor is not yet created, or is being reconstructed, cache the request
+    // and submit after actor is alive.
+    // TODO(zhijunfu): it might be possible for a user to specify an invalid
+    // actor handle (e.g. from unpickling), in that case it might be desirable
+    // to have a timeout to mark it as invalid if it doesn't show up in the
+    // specified time.
+    pending_requests_[actor_id].emplace_back(std::move(request));
+    return Status::OK();
+  } else if (iter->second.state_ == ActorTableData::ALIVE) {
+    // Actor is alive, submit the request.
+    if (rpc_clients_.count(actor_id) == 0) {
+      // If rpc client is not available, then create it.
+      ConnectAndSendPendingTasks(actor_id, iter->second.location_.first,
+                                 iter->second.location_.second);
+    }
+
+    // Submit request.
+    auto &client = rpc_clients_[actor_id];
+    return PushTask(*client, *request, task_id, num_returns);
+  } else {
+    // Actor is dead, treat the task as failure.
+    RAY_CHECK(iter->second.state_ == ActorTableData::DEAD);
+    TreatTaskAsFailed(task_id, num_returns, rpc::ErrorType::ACTOR_DIED);
+    return Status::IOError("Actor is dead.");
+  }
+}
+
+Status CoreWorkerDirectActorTaskSubmitter::SubscribeActorUpdates() {
+  // Register a callback to handle actor notifications.
+  auto actor_notification_callback = [this](const ActorID &actor_id,
+                                            const ActorTableData &actor_data) {
+    std::unique_lock<std::mutex> guard(mutex_);
+    actor_states_.erase(actor_id);
+    actor_states_.emplace(
+        actor_id,
+        ActorStateData(actor_data.state(), actor_data.ip_address(), actor_data.port()));
+
+    if (actor_data.state() == ActorTableData::ALIVE) {
+      // Check if this actor is the one that we're interested, if we already have
+      // a connection to the actor, or have pending requests for it, we should
+      // create a new connection.
+      if (pending_requests_.count(actor_id) > 0) {
+        ConnectAndSendPendingTasks(actor_id, actor_data.ip_address(), actor_data.port());
+      }
+    } else {
+      // Remove rpc client if it's dead or being reconstructed.
+      rpc_clients_.erase(actor_id);
+    }
+
+    RAY_LOG(INFO) << "received notification on actor, state="
+                  << static_cast<int>(actor_data.state()) << ", actor_id: " << actor_id
+                  << ", ip address: " << actor_data.ip_address()
+                  << ", port: " << actor_data.port();
+  };
+
+  return gcs_client_.Actors().AsyncSubscribe(actor_notification_callback, nullptr);
+}
+
+void CoreWorkerDirectActorTaskSubmitter::ConnectAndSendPendingTasks(
+    const ActorID &actor_id, std::string ip_address, int port) {
+  std::unique_ptr<rpc::DirectActorClient> grpc_client(
+      new rpc::DirectActorClient(ip_address, port, client_call_manager_));
+  // replace old rpc client if it exists.
+  rpc_clients_[actor_id] = std::move(grpc_client);
```

Comment:
use emplace? because the old client should have been removed before

Reply:
ok